src/qibo/backends/_clifford_operations.py

```diff
@@ -0,0 +1,449 @@
+    )
+    x_target_q = x[:, control_q] ^ x[:, target_q]
+    z_control_q = z[:, control_q] ^ z[:, target_q] ^ x[:, target_q]
+    z_target_q = z[:, target_q] ^ x[:, control_q]
+    symplectic_matrix[:-1, target_q] = x_target_q
+    symplectic_matrix[:-1, nqubits + control_q] = z_control_q
+    symplectic_matrix[:-1, nqubits + target_q] = z_target_q
+    return symplectic_matrix
+
+
+def CRX(symplectic_matrix, control_q, target_q, nqubits, theta):
+    # theta = 4 * n * pi
+    if theta % (4 * np.pi) == 0:
+        return I(symplectic_matrix, target_q, nqubits)
+    # theta = pi + 4 * n * pi
+    elif (theta / np.pi - 1) % 4 == 0:
+        symplectic_matrix = X(symplectic_matrix, target_q, nqubits)
+        symplectic_matrix = CZ(symplectic_matrix, control_q, target_q, nqubits)
+        symplectic_matrix = X(symplectic_matrix, target_q, nqubits)
+        return CY(symplectic_matrix, control_q, target_q, nqubits)
+    # theta = 2 * pi + 4 * n * pi
+    elif (theta / (2 * np.pi) - 1) % 2 == 0:
+        symplectic_matrix = CZ(symplectic_matrix, control_q, target_q, nqubits)
+        symplectic_matrix = Y(symplectic_matrix, target_q, nqubits)
+        symplectic_matrix = CZ(symplectic_matrix, control_q, target_q, nqubits)
+        return Y(symplectic_matrix, target_q, nqubits)
+    # theta = 3 * pi + 4 * n * pi
+    elif (theta / np.pi - 3) % 4 == 0:
+        symplectic_matrix = X(symplectic_matrix, target_q, nqubits)
+        symplectic_matrix = CY(symplectic_matrix, control_q, target_q, nqubits)
+        symplectic_matrix = X(symplectic_matrix, target_q, nqubits)
+        return CZ(symplectic_matrix, control_q, target_q, nqubits)
+
+
+def CRZ(symplectic_matrix, control_q, target_q, nqubits, theta):
+    # theta = 4 * n * pi
+    if theta % (4 * np.pi) == 0:
+        return I(symplectic_matrix, target_q, nqubits)
+    # theta = pi + 4 * n * pi
+    elif (theta / np.pi - 1) % 4 == 0:
+        symplectic_matrix = X(symplectic_matrix, target_q, nqubits)
+        symplectic_matrix = CY(symplectic_matrix, control_q, target_q, nqubits)
+        symplectic_matrix = X(symplectic_matrix, target_q, nqubits)
+        return CNOT(symplectic_matrix, control_q, target_q, nqubits)
+    # theta = 2 * pi + 4 * n * pi
+    elif (theta / (2 * np.pi) - 1) % 2 == 0:
+        symplectic_matrix = CZ(symplectic_matrix, control_q, target_q, nqubits)
+        symplectic_matrix = X(symplectic_matrix, target_q, nqubits)
+        symplectic_matrix = CZ(symplectic_matrix, control_q, target_q, nqubits)
+        return X(symplectic_matrix, target_q, nqubits)
+    # theta = 3 * pi + 4 * n * pi
+    elif (theta / np.pi - 3) % 4 == 0:
+        symplectic_matrix = CNOT(symplectic_matrix, control_q, target_q, nqubits)
+        symplectic_matrix = X(symplectic_matrix, target_q, nqubits)
+        symplectic_matrix = CY(symplectic_matrix, control_q, target_q, nqubits)
+        return X(symplectic_matrix, target_q, nqubits)
```

Comment:
You know that if no condition is satisfied, the function will just return `None`?

I'd rather raise an error...

Reply:
Moreover, this is another repeated pattern. It would really be nice to have a syntax for gates composition, to make it something like:
```suggestion
    theta_switch(
        theta,
        I,
        X | CY | X | CNOT,
        CZ | X | CZ | X,
        CNOT | X | CY | X,
        symplectic_matrix, control_q, target_q, nqubits
    )
```
with `theta_switch` a function applying the repeated pattern (and I chose `|` at random, any operator would be fine).

If you're interested in, we could try to design something like that.

Reply:
I am not really familiar with monads, but sounds like an interesting idea!

Reply:
> I am not really familiar with monads, but sounds like an interesting idea!

That was kind of a joke: there are no monads in Python. With monads, you can manage a state in a pure functional way, without explicitly passing it in every function invocation, since it will be part of the type (it's slightly more complicate actually https://learnyouahaskell.com/a-fistful-of-monads...)

In this case, the idea is that your state is exactly the symplectic matrix, and you'd like to make operations on that.

However, we don't need such a fancy thing (though it would be kind of elegant), but here we just need a way to batch some operations and execute them.
We can just make the operations instances of a class defining a `__or__` containing the function, something like:

```py
@dataclass
class CliffordOp:
    ops: list[Callable]

    def __or__(self, other):
        ...

    def __call__(self, state):
        for f in ops:
            state = f(state)
        return state

    @classmethod
    def atom(cls, f):
        return cls([f])

def _CNOT(state):
    ...

CNOT = CliffordOp.atom(_CNOT)
```

The details could be improved, but that's the idea.

Another option is to just compose the function in `__or__`:
```py
@dataclass
class CliffordOp:
    f:  Callable

    def __or__(self, other):
        cls = type(self)
        return cls(lambda state: self(other(state)))

    def __call__(self, state):
        return self.f(state)

def _CNOT(state):
    ...

CNOT = CliffordOp(_CNOT)
```
but Python is not a functional language, so it has limited support for optimizations on calls, I'm not sure how much I'd play this game of nesting...

Reply:
@BrunoLiegiBastonLiegi this we decided to postpone anyhow, so you could open an issue and resolve the conversation.